File: app/src/level.py
# Copyright (C) 2025 Logical Mechanism LLC
# SPDX-License-Identifier: GPL-3.0-only
from src.files import save_json


def full_level_to_file(r1b: str, r2_g1b: str, r2_g2b: str, r4b: str) -> None:
    """
    Write a "full-level" entry JSON artifact to disk.

    A "full-level" entry includes:
    - `r1b`: a G1 element (bytes/hex string)
    - `r2`: a sum-type structure containing:
        - `r2_g1b`: the G1 component
        - `r2_g2b`: the G2 component (wrapped in a constructor)
    - `r4b`: an additional G1 element (bytes/hex string)

    The output schema matches a Plutus/Aiken constructor encoding. In particular,
    the nested structure corresponds to a variant where the second field has a
    populated G2 sub-field (i.e., not the "empty" constructor used by half-level).

    Output path (fixed):
        ../data/full-level.json

    Args:
        r1b: Hex/bytes string for the `r1` point (G1).
        r2_g1b: Hex/bytes string for the `r2` G1 component.
        r2_g2b: Hex/bytes string for the `r2` G2 component.
        r4b: Hex/bytes string for the `r4` point (G1).

    Returns:
        None. Writes the JSON artifact via `save_json`.

    Raises:
        Any exceptions raised by `save_json` (e.g., invalid path or permissions)
        will propagate.
    """
    path = "../data/full-level.json"
    data = {
        "constructor": 0,
        "fields": [
            {"bytes": r1b},
            {
                "constructor": 0,
                "fields": [
                    {"bytes": r2_g1b},
                    {
                        "constructor": 0,
                        "fields": [{"bytes": r2_g2b}],
                    },
                ],
            },
            {"bytes": r4b},
        ],
    }
    save_json(path, data)


def half_level_to_file(r1b: str, r2_g1b: str, r4b: str) -> None:
    """
    Write a "half-level" entry JSON artifact to disk.

    A "half-level" entry includes:
    - `r1b`: a G1 element (bytes/hex string)
    - `r2`: a sum-type structure containing:
        - `r2_g1b`: the G1 component
        - an *empty* alternative (constructor=1) indicating the G2 component
          is not present yet
    - `r4b`: an additional G1 element (bytes/hex string)

    This represents an entry before a subsequent hop fills in the G2 component
    (which is what `full_level_to_file` encodes).

    Output path (fixed):
        ../data/half-level.json

    Args:
        r1b: Hex/bytes string for the `r1` point (G1).
        r2_g1b: Hex/bytes string for the `r2` G1 component.
        r4b: Hex/bytes string for the `r4` point (G1).

    Returns:
        None. Writes the JSON artifact via `save_json`.

    Raises:
        Any exceptions raised by `save_json` (e.g., invalid path or permissions)
        will propagate.
    """
    path = "../data/half-level.json"
    data = {
        "constructor": 0,
        "fields": [
            {"bytes": r1b},
            {
                "constructor": 0,
                "fields": [
                    {"bytes": r2_g1b},
                    {
                        "constructor": 1,
                        "fields": [],
                    },
                ],
            },
            {"bytes": r4b},
        ],
    }
    save_json(path, data)

File: app/ui/fe/src/services/crypto/level.ts
/**
 * Encryption Level Structures
 *
 * Ported from Python: src/level.py
 * Represents half-level and full-level encryption entries.
 */

/**
 * Half-level encryption entry (before re-encryption hop).
 */
export interface HalfLevel {
  r1: string; // G1 point (96 hex chars)
  r2_g1: string; // G1 component (96 hex chars)
  r4: string; // G2 point (192 hex chars) - commitment computed from H1, H2, H3
}

/**
 * Full-level encryption entry (after re-encryption hop).
 */
export interface FullLevel {
  r1: string; // G1 point (96 hex chars)
  r2_g1: string; // G1 component (96 hex chars)
  r2_g2: string; // G2 component (192 hex chars)
  r4: string; // G2 point (192 hex chars) - commitment computed from H1, H2, H3
}

/**
 * Convert a HalfLevel to Plutus/Aiken JSON format.
 */
export function halfLevelToPlutusJson(level: HalfLevel): object {
  return {
    constructor: 0,
    fields: [{ bytes: level.r1 }, { bytes: level.r2_g1 }, { bytes: level.r4 }],
  };
}

/**
 * Convert a FullLevel to Plutus/Aiken JSON format.
 * Uses constructor 0 with nested structure for the G2 component.
 */
export function fullLevelToPlutusJson(level: FullLevel): object {
  return {
    constructor: 0,
    fields: [
      {
        constructor: 0,
        fields: [
          { bytes: level.r1 },
          { bytes: level.r2_g1 },
          { bytes: level.r2_g2 },
          { bytes: level.r4 },
        ],
      },
    ],
  };
}

/**
 * Create an empty full-level JSON (constructor 1 - Nothing variant).
 */
export function emptyFullLevelToPlutusJson(): object {
  return {
    constructor: 1,
    fields: [],
  };
}
